Cassandra integration tests are failing consistently

<!-- **Are you in the right place?**
1. For issues or feature requests, please create an issue in this repository.
2. For general technical and non-technical questions, we are happy to help you on our [Rook.io Slack](https://slack.rook.io/).
3. Did you already search the existing open issues for anything similar? -->

**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
The Cassandra integration tests are failing since yesterday in `master` and `release-1.6` branches. For example, in [this release build](https://jenkins.rook.io/blue/rest/organizations/jenkins/pipelines/rook/pipelines/rook/branches/release-1.6/runs/6/nodes/51/steps/131/log/?start=0), the failure shows:
```
[2021-04-08T19:55:05.422Z] === RUN   TestCassandraSuite/TestCassandraClusterCreation
[2021-04-08T19:55:05.422Z] 2021-04-08 19:55:04.508376 I | integrationTest: Verifying that all expected pods in cassandra cluster cassandra-ns are running
[2021-04-08T19:55:05.422Z] 2021-04-08 19:55:04.510359 I | testutil: 1 of 1 pods with label app=rook-cassandra-operator were found
[2021-04-08T19:55:37.505Z] 2021-04-08 19:55:34.515933 I | testutil: 1 of 1 pods with label app=rook-cassandra were found
[2021-04-08T19:55:37.505Z] 2021-04-08 19:55:34.522604 D | exec: Running command: kubectl exec -n cassandra-ns cassandra-ns-us-east-1-us-east-1a-0 -- hostname -i
[2021-04-08T19:56:09.562Z] 2021-04-08 19:56:04.873404 D | exec: Running command: kubectl exec -n cassandra-ns cassandra-ns-us-east-1-us-east-1a-0 -- cqlsh -e 
[2021-04-08T19:56:09.563Z] CREATE KEYSPACE IF NOT EXISTS test WITH REPLICATION = {
[2021-04-08T19:56:09.563Z] 'class': 'SimpleStrategy',
[2021-04-08T19:56:09.563Z] 'replication_factor': 1
[2021-04-08T19:56:09.563Z] };
[2021-04-08T19:56:09.563Z] USE test;
[2021-04-08T19:56:09.563Z] CREATE TABLE IF NOT EXISTS map (key text, value text, PRIMARY KEY(key));
[2021-04-08T19:56:09.563Z] INSERT INTO map (key, value) VALUES('test_key', 'test_value');
[2021-04-08T19:56:09.563Z] SELECT key,value FROM map WHERE key='test_key'; Defaulted container "cassandra" out of: cassandra, rook-install (init)
[2021-04-08T19:56:09.563Z] 10.32.0.6
[2021-04-08T19:56:09.563Z] 2021-04-08 19:56:05.436749 E | utils: Failed to execute: kubectl [exec -n cassandra-ns cassandra-ns-us-east-1-us-east-1a-0 -- cqlsh -e 
[2021-04-08T19:56:09.563Z] CREATE KEYSPACE IF NOT EXISTS test WITH REPLICATION = {
[2021-04-08T19:56:09.563Z] 'class': 'SimpleStrategy',
[2021-04-08T19:56:09.563Z] 'replication_factor': 1
[2021-04-08T19:56:09.563Z] };
[2021-04-08T19:56:09.563Z] USE test;
[2021-04-08T19:56:09.563Z] CREATE TABLE IF NOT EXISTS map (key text, value text, PRIMARY KEY(key));
[2021-04-08T19:56:09.563Z] INSERT INTO map (key, value) VALUES('test_key', 'test_value');
[2021-04-08T19:56:09.563Z] SELECT key,value FROM map WHERE key='test_key'; Defaulted container "cassandra" out of: cassandra, rook-install (init)
[2021-04-08T19:56:09.563Z] 10.32.0.6] : exit status 1. Defaulted container "cassandra" out of: cassandra, rook-install (init)
[2021-04-08T19:56:09.563Z] Traceback (most recent call last):
[2021-04-08T19:56:09.563Z]   File "/opt/cassandra/bin/cqlsh.py", line 2443, in <module>
[2021-04-08T19:56:09.563Z]     main(*read_options(sys.argv[1:], os.environ))
[2021-04-08T19:56:09.563Z]   File "/opt/cassandra/bin/cqlsh.py", line 2421, in main
[2021-04-08T19:56:09.563Z]     encoding=options.encoding)
[2021-04-08T19:56:09.563Z]   File "/opt/cassandra/bin/cqlsh.py", line 485, in __init__
[2021-04-08T19:56:09.563Z]     load_balancing_policy=WhiteListRoundRobinPolicy([self.hostname]),
[2021-04-08T19:56:09.563Z]   File "/opt/cassandra/bin/../lib/cassandra-driver-internal-only-3.11.0-bb96859b.zip/cassandra-driver-3.11.0-bb96859b/cassandra/policies.py", line 417, in __init__
[2021-04-08T19:56:09.563Z] socket.gaierror: [Errno -2] Name or service not known
[2021-04-08T19:56:09.563Z] command terminated with exit code 1
[2021-04-08T19:56:09.563Z] 2021-04-08 19:56:05.436811 I | integrationTest: cassandra cql command exited, err: kubectl exec command cqlsh failed on pod cassandra-ns-us-east-1-us-east-1a-0 in namespace cassandra-ns. Failed to run: kubectl [exec -n cassandra-ns cassandra-ns-us-east-1-us-east-1a-0 -- cqlsh -e 
```

After a number of retries, it fails here:
```
[2021-04-08T19:58:51.398Z]     z_cassandra_test.go:175: 
[2021-04-08T19:58:51.398Z]         	Error Trace:	z_cassandra_test.go:175
[2021-04-08T19:58:51.398Z]         	            				z_cassandra_test.go:107
[2021-04-08T19:58:51.398Z]         	Error:      	Received unexpected error:
[2021-04-08T19:58:51.398Z]         	            	kubectl exec command cqlsh failed on pod cassandra-ns-us-east-1-us-east-1a-0 in namespace cassandra-ns. Failed to run: kubectl [exec -n cassandra-ns cassandra-ns-us-east-1-us-east-1a-0 -- cqlsh -e 
[2021-04-08T19:58:51.398Z]         	            	CREATE KEYSPACE IF NOT EXISTS test WITH REPLICATION = {
[2021-04-08T19:58:51.398Z]         	            	'class': 'SimpleStrategy',
[2021-04-08T19:58:51.398Z]         	            	'replication_factor': 1
[2021-04-08T19:58:51.398Z]         	            	};
[2021-04-08T19:58:51.398Z]         	            	USE test;
[2021-04-08T19:58:51.398Z]         	            	CREATE TABLE IF NOT EXISTS map (key text, value text, PRIMARY KEY(key));
[2021-04-08T19:58:51.398Z]         	            	INSERT INTO map (key, value) VALUES('test_key', 'test_value');
[2021-04-08T19:58:51.398Z]         	            	SELECT key,value FROM map WHERE key='test_key'; Defaulted container "cassandra" out of: cassandra, rook-install (init)
[2021-04-08T19:58:51.398Z]         	            	10.32.0.6] : exit status 1
[2021-04-08T19:58:51.398Z]         	Test:       	TestCassandraSuite/TestCassandraClusterCreation
[2021-04-08T19:58:51.398Z]     z_cassandra_test.go:176: 
[2021-04-08T19:58:51.398Z]         	Error Trace:	z_cassandra_test.go:176
[2021-04-08T19:58:51.398Z]         	            				z_cassandra_test.go:107
[2021-04-08T19:58:51.398Z]         	Error:      	Should be true
[2021-04-08T19:58:51.398Z]         	Test:       	TestCassandraSuite/TestCassandraClusterCreation
[2021-04-08T19:58:51.398Z]     z_cassandra_test.go:177: 
[2021-04-08T19:58:51.398Z]         	Error Trace:	z_cassandra_test.go:177
[2021-04-08T19:58:51.398Z]         	            				z_cassandra_test.go:107
[2021-04-08T19:58:51.398Z]         	Error:      	Should be true
[2021-04-08T19:58:51.398Z]         	Test:       	TestCassandraSuite/TestCassandraClusterCreation
[2021-04-08T19:58:51.398Z] --- FAIL: TestCassandraSuite (248.02s)
[2021-04-08T19:58:51.398Z]     --- FAIL: TestCassandraSuite/TestCassandraClusterCreation (226.54s)
```

This does not appear to be related to a change in Rook.

**Expected behavior:**
Integration tests should pass

**How to reproduce it (minimal and precise):**
<!-- Please let us know any circumstances for reproduction of your bug. -->
Run the Cassandra integration tests

--- a/.github/workflows/integration-test-cassandra-suite.yaml
+++ b/.github/workflows/integration-test-cassandra-suite.yaml
@@ -1,0 +1,113 @@
+name: Integration test CassandraSuite
+on:
+  push:
+    tags:
+      - v*
+    branches:
+      - master
+      - release-*
+  pull_request:
+
+jobs:
+  TestCassandraSuite:
+    if: ${{ github.event_name == 'pull_request' && github.ref != 'refs/heads/master' }}
+    runs-on: ubuntu-18.04
+    strategy:
+      fail-fast: false
+      matrix:
+        kubernetes-versions : ['v1.15.12', 'v1.21.0']
+    steps:
+    - name: checkout
+      uses: actions/checkout@v2
+
+    - name: setup golang
+      uses: actions/setup-go@v2
+      with:
+        go-version: 1.15
+
+    - name: setup minikube
+      uses: manusa/actions-setup-minikube@v2.3.1
+      with:
+        minikube version: 'v1.18.1'
+        kubernetes version: ${{ matrix.kubernetes-versions }}
+        start args: --memory 6g --cpus=2
+        github token: ${{ secrets.GITHUB_TOKEN }}
+
+    - name: check k8s cluster status
+      run: |
+        kubectl cluster-info
+        kubectl get pods -n kube-system
+
+    - name: build rook
+      run: |
+        # set VERSION to a dummy value since Jenkins normally sets it for us. Do this to make Helm happy and not fail with "Error: Invalid Semantic Version"
+        GOPATH=$(go env GOPATH) make clean && make -j$nproc IMAGES='cassandra' VERSION=0 build
+        docker images
+        docker tag $(docker images|awk '/build-/ {print $1}') rook/cassandra:master
+
+    - name: TestCassandraSuite
+      run: |
+        go test -v -timeout 1800s -run CassandraSuite github.com/rook/rook/tests/integration
+
+    - name: Artifact
+      uses: actions/upload-artifact@v2
+      if: failure()
+      with:
+        name: cassandra-suite-artifact
+        path: /home/runner/work/rook/rook/tests/integration/_output/tests/
+
+    - name: setup tmate session for debugging
+      if: failure() && contains(github.event.pull_request.labels.*.name, 'debug-cassandra-suite')
+      uses: mxschmitt/action-tmate@v3
+
+  TestCassandraSuite-Master:
+    if: github.ref == 'refs/heads/master' || github.ref == 'refs/tags/release-*'
+    runs-on: ubuntu-18.04
+    strategy:
+      fail-fast: false
+      matrix:
+        kubernetes-versions : ['v1.15.12','v1.18.15','v1.20.5','v1.21.0']
+    steps:
+    - name: checkout
+      uses: actions/checkout@v2
+
+    - name: setup golang
+      uses: actions/setup-go@v2
+      with:
+        go-version: 1.15
+
+    - name: setup minikube
+      uses: manusa/actions-setup-minikube@v2.3.1
+      with:
+        minikube version: 'v1.18.1'
+        kubernetes version: ${{ matrix.kubernetes-versions }}
+        start args: --memory 6g --cpus=2
+        github token: ${{ secrets.GITHUB_TOKEN }}
+
+    - name: check k8s cluster status
+      run: |
+        kubectl cluster-info
+        kubectl get pods -n kube-system
+
+    - name: build rook
+      run: |
+        # set VERSION to a dummy value since Jenkins normally sets it for us. Do this to make Helm happy and not fail with "Error: Invalid Semantic Version"
+        GOPATH=$(go env GOPATH) make clean && make -j$nproc IMAGES='cassandra' VERSION=0 build
+        docker images
+        docker tag $(docker images|awk '/build-/ {print $1}') rook/cassandra:master
+
+    - name: TestCassandraSuite
+      run: |
+       export DEVICE_FILTER=$(lsblk|awk '/14G/ {print $1}'| head -1)
+       SKIP_CLEANUP_POLICY=false go test -v -timeout 1800s -run CassandraSuite github.com/rook/rook/tests/integration
+
+    - name: Artifact
+      uses: actions/upload-artifact@v2
+      if: failure()
+      with:
+        name: cassandra-suite-artifact
+        path: /home/runner/work/rook/rook/tests/integration/_output/tests/
+
+    - name: setup tmate session for debugging
+      if: failure() && contains(github.event.pull_request.labels.*.name, 'cassandra-suite')
+      uses: mxschmitt/action-tmate@v3

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -234,6 +234,7 @@ def RunIntegrationTest(k, v) {
                         sh '''#!/bin/bash
                               set -o pipefail
                               export KUBECONFIG=$HOME/admin.conf \
+                                  SKIP_CASSANDRA_TESTS=true \
                                   TEST_ENV_NAME='''+"${k}"+''' \
                                   TEST_BASE_DIR="WORKING_DIR" \
                                   TEST_LOG_COLLECTION_LEVEL='''+"${env.getLogs}"+''' \

--- a/tests/framework/installer/cassandra_manifests.go
+++ b/tests/framework/installer/cassandra_manifests.go
@@ -209,6 +209,8 @@ rules:
     verbs:
       - get
       - list
+      - patch
+      - watch
   - apiGroups:
       - cassandra.rook.io
     resources:

--- a/tests/integration/z_cassandra_test.go
+++ b/tests/integration/z_cassandra_test.go
@@ -18,7 +18,6 @@ package integration
 
 import (
 	"context"
-	"fmt"
 	"os"
 	"strings"
 	"testing"
@@ -30,6 +29,7 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"github.com/stretchr/testify/suite"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // ************************************************
@@ -132,14 +132,58 @@ func (s *CassandraSuite) Teardown() {
 // and CQL is working.
 func (s *CassandraSuite) CheckClusterHealth() {
 	// Verify that cassandra-operator is running
-	logger.Infof("Verifying that all expected pods in cassandra cluster %s are running", s.namespace)
-	assert.True(s.T(), s.k8sHelper.CheckPodCountAndState("rook-cassandra-operator", s.systemNamespace, 1, "Running"), "rook-cassandra-operator must be in Running state")
+	operatorName := "rook-cassandra-operator"
+	logger.Infof("Verifying that all expected pods of cassandra operator are ready")
+	ready := utils.Retry(10, 30*time.Second,
+		"Waiting for Cassandra operator to be ready", func() bool {
+			sts, err := s.k8sHelper.Clientset.AppsV1().StatefulSets(s.systemNamespace).Get(context.TODO(), operatorName, v1.GetOptions{})
+			if err != nil {
+				logger.Errorf("Error getting Cassandra operator `%s`", operatorName)
+				return false
+			}
+			if sts.Generation != sts.Status.ObservedGeneration {
+				logger.Infof("Operator Statefulset has not converged yet")
+				return false
+			}
+			if sts.Status.UpdatedReplicas != *sts.Spec.Replicas {
+				logger.Error("Operator StatefulSet is rolling updating")
+				return false
+			}
+			if sts.Status.ReadyReplicas != *sts.Spec.Replicas {
+				logger.Infof("Statefulset not ready. Got: %v, Want: %v",
+					sts.Status.ReadyReplicas, sts.Spec.Replicas)
+				return false
+			}
+			return true
+		})
+	assert.True(s.T(), ready, "Timed out waiting for Cassandra operator to become ready")
 
-	// Give the StatefulSet a head start
-	// CheckPodCountAndState timeout might be too fast and the test may fail
-	time.Sleep(30 * time.Second)
 	// Verify cassandra cluster instances are running OK
-	assert.True(s.T(), s.k8sHelper.CheckPodCountAndState("rook-cassandra", s.namespace, s.instanceCount, "Running"), fmt.Sprintf("%d rook-cassandra pods must be in running state", s.instanceCount))
+	clusterName := "cassandra-ns"
+	clusterNamespace := "cassandra-ns"
+	ready = utils.Retry(10, 30*time.Second,
+		"Waiting for Cassandra cluster to be ready", func() bool {
+			c, err := s.k8sHelper.RookClientset.CassandraV1alpha1().Clusters(clusterNamespace).Get(context.TODO(), clusterName, v1.GetOptions{})
+			if err != nil {
+				logger.Errorf("Error getting Cassandra cluster `%s`", clusterName)
+				return false
+			}
+			for rackName, rack := range c.Status.Racks {
+				var desiredMembers int32
+				for _, r := range c.Spec.Datacenter.Racks {
+					if r.Name == rackName {
+						desiredMembers = r.Members
+						break
+					}
+				}
+				if !(desiredMembers == rack.Members && rack.Members == rack.ReadyMembers) {
+					logger.Infof("Rack `%s` is not ready yet", rackName)
+					return false
+				}
+			}
+			return true
+		})
+	assert.True(s.T(), ready, "Timed out waiting for Cassandra cluster to become ready")
 
 	// Determine a pod name for the cluster
 	podName := "cassandra-ns-us-east-1-us-east-1a-0"

--- a/tests/integration/z_cassandra_test.go
+++ b/tests/integration/z_cassandra_test.go
@@ -17,7 +17,9 @@ limitations under the License.
 package integration
 
 import (
+	"context"
 	"fmt"
+	"os"
 	"strings"
 	"testing"
 	"time"
@@ -52,6 +54,9 @@ type CassandraSuite struct {
 // TestCassandraSuite initiates the CassandraSuite
 func TestCassandraSuite(t *testing.T) {
 	if installer.SkipTestSuite(installer.CassandraTestSuite) {
+		t.Skip()
+	}
+	if os.Getenv("SKIP_CASSANDRA_TESTS") == "true" {
 		t.Skip()
 	}
 

